### PR TITLE
Interrupt refactor

### PR DESCRIPTION
**Pull request by steveklabnik**
This is a big refactoring of the way interrupts are handled. I took some inspiration from @phil-opp, @emk, Redox, (specifically, @ticki) and an `inb`/`outb` combo from @thepowersgang. All different bits of their OSes' code and several OSDev/OSDevr tutorials. Whew!

These commits should probably be cleaned up, and this isn't even quite in a final state; I should be using a `lazy_static` rather than making things `Copy`, and I would like to move a lot of these home-grown structures to the `x86` crate instead. But before I do, I wanted to have written it all myself. Ha!

Thoughts, all?

### CODE REVIEW

File: src/interrupts/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,190 @@
+#![feature(const_fn)]
+#![feature(asm)]
+#![no_std]
+
+#[macro_use]
+extern crate vga;
+
+extern {
+    fn isr0();
```
Comment (eddyb) [inline src/interrupts/src/lib.rs]: But... naked functions :laughing:

**Comment by ticki** [comment]
Awesome work! It seems that you don't use naked functions? Why not?

**Comment by steveklabnik** [comment]
@ticki @eddyb I should. this code originally landed since before naked functions had existed, but now that they do, i should make use of them.

**Comment by WilsonGiese** [comment]
I've been following along with your development and it's looking great to me! Excited for the upcoming work on the kernel 👍

**Comment by steveklabnik** [comment]
Merging this for now, because it's still an improvement. I'll open issues for the other stuff.

**Pull request merged**